app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,181 @@
+import javax.crypto.KeyGenerator
+import javax.crypto.NoSuchPaddingException
+import javax.crypto.SecretKey
+
+@RequiresApi(Build.VERSION_CODES.M)
+class BiometricAuthenticationHandler(
+        private val context: Context) : FingerprintManagerCompat.AuthenticationCallback(), LifecycleObserver {
+
+    private var fingerprintManager = FingerprintManagerCompat.from(context)
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+    private var keyStore: KeyStore? = null
+    private var keyGenerator: KeyGenerator? = null
+    private var cryptoObject: FingerprintManagerCompat.CryptoObject? = null
+
+    var needsAuth = false; private set
+    var biometricFragment: BiometricAuthenticationDialogFragment? = null; private set
+
+    init {
+        try {
+            keyStore = KeyStore.getInstance("AndroidKeyStore")
+
+        } catch (err: KeyStoreException) {
+            throw RuntimeException("Failed to get instance of KeyStore", err)
+        }
+
+        try {
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchProviderException ->
+                    throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+                else -> throw err
+            }
+        }
+
+        createKey(DEFAULT_KEY_NAME, false)
+
+        val defaultCipher: Cipher
+        try {
+            defaultCipher = Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/"
+                    + KeyProperties.BLOCK_MODE_CBC + "/"
+                    + KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchPaddingException ->
+                    throw RuntimeException("Failed to get an instance of Cipher", err)
+                else -> throw err
+            }
+        }
+
+        if (initCipher(defaultCipher, DEFAULT_KEY_NAME)) {
+            cryptoObject = FingerprintManagerCompat.CryptoObject(defaultCipher)
+        }
+
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private fun initCipher(cipher: Cipher, keyName: String): Boolean {
+        try {
+            keyStore?.load(null)
+            val key = keyStore?.getKey(keyName, null) as SecretKey
+            cipher.init(Cipher.ENCRYPT_MODE, key)
+            return true
+        }  catch (err: Exception) {
+            when (err) {
+                is KeyPermanentlyInvalidatedException -> return false
+                is KeyStoreException,
+                is CertificateException,
+                is UnrecoverableKeyException,
+                is IOException,
+                is NoSuchAlgorithmException,
+                is InvalidKeyException -> throw RuntimeException("Failed to init cipher", err)
```

Comment:
\[detekt\] ERROR: \[exceptions/ThrowRuntimeException\] Caught exception is too generic\. Prefer catching specific Exceptions to handle specific error cases\.